rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,345 @@
+
+The following might not be implemented before 2022-03-01T00:00Z,
+but the above will still regardless be required at that point in time.
+
+## Verifying flake Git inputs
+
+### Updating flake Git inputs
+[updating]: #updating-flake-git-inputs
+
+The design is inspired by what is described here: https://guix.gnu.org/en/blog/2020/securing-updates/.
+
+When updating any existing flake input of type `git` from an old commit, Nix must verify that
+the new commit is trusted.
+
+The old commit is always trusted.
+
+In the context of Git, a parent is a direct parent in the DAG of commits.
+
+If a commit has a trusted parent, then the commit is trusted if
+it doesn't have a trusted parent with `/.authorized-committers.nix`,
+
+If a commit has a trusted parent,
+and is signed by a committer listed in each of the trusted
+parents' `/.authorized-committers.nix`, then the commit is trusted.
+
+#### The format of `/.authorized-committers.nix`
+
+An example: `/.authorized-committers.nix`:
+```nix
+{
+  committers = {
+    jacob = "XXXX XXXX XXXX XXXX XXXX  XXXX XXXX XXXX XXXX XXXX";
+    jonathan = "YYYY YYYY YYYY YYYY YYYY  YYYY YYYY YYYY YYYY YYYY";
+  };
+}
+```
+
+Specifically, it will contain a Nix expression, that evaluates
+to a set with a single attribute `committers`, which contains
+another set where names are mapped to PGP fingerprints. This is a set
+instead of a list to make it easy to get the fingerprint for a specific
+person.
+
+### New flake Git inputs
+
+When a flake input is initialized, there is no good way to check who to trust
+automatically, i.e. it is [Trust on First Use](https://en.wikipedia.org/wiki/Trust_on_first_use).
+
+To secure initializations, new attributes are added to the `git` input type:
+- `startsAt` (optional): This is the first commit to use, therefore verification
+  starts from this commit on.
+
+The standard registry entries will have the `startsAt` attribute
+added, to ensure that NixOS users fetch trusted source code.
+
+### Other new input attributes
+
+- `insecureUpdate` (optional): If set to true, then the flake will update without errors
+  to untrusted commits, however it will still emit a warning.
+- `alwaysTrust` (optional): Commits signed by these PGP fingerprints will always be trusted.
+
+## Speeding up fetching Git
+
+Cryptographically verifying Git commits is easy, but
+Git is currently often not suitable for using as a flake input.
+Nixpkgs is growing at an extreme rate, and with each
+revision cloning it becomes slower.
+For the author, a full clone from scratch takes over 4 minutes.
+
+To avoid this problem with the Git protocol, Nix currently
+supports the `github` type, which fetches archives from GitHub,
+and updates the flake input using proprietary GitHub APIs.
+This is GitHub specific, isn't usable on other platforms, and can
+not take advantage of the above method for verifying source code
+integrity.
+
+We can however speed up fetching Git repositories heavily
```

Comment:
More stuff that isn't specific to flakes but could as well be done using any builtin git fetcher.

Reply:
Yes this isn't specific to flakes at all. This isn't core to the RFC, it's more so an explanation for how fetching Nixpkgs through Git will be done without taking ages.